### PR TITLE
Adding cuda-nvtx recipe

### PR DESCRIPTION
**Pull request by adibbley**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/cuda-nvtx/meta.yaml
Comment (bdice) [inline recipes/cuda-nvtx/meta.yaml:40]: This recipe needs pinnings to `cuda-version` on each of the subpackages below.
```suggestion
  - name: cuda-nvtx
    requirements:
      run:
        - cuda-version =12.0
```

File: recipes/cuda-nvtx/meta.yaml
Comment (bdice) [inline recipes/cuda-nvtx/meta.yaml:54]:
```suggestion
        - cuda-nvtx-dev_{{ target_platform }} =={{ version }}
        - cuda-version =12.0
```

File: recipes/cuda-nvtx/meta.yaml
Comment (bdice) [inline recipes/cuda-nvtx/meta.yaml:65]:
```suggestion
      - {{ pin_subpackage("cuda-nvtx", max_pin='x') }}
    requirements:
      run:
        - cuda-version =12.0
```

File: recipes/cuda-nvtx/meta.yaml
Comment (bdice) [inline recipes/cuda-nvtx/meta.yaml:2]:
```suggestion
{% set version = "12.0.76" %}
{% set cuda_version = "12.0" %}
```

File: recipes/cuda-nvtx/meta.yaml
Comment (bdice) [inline recipes/cuda-nvtx/meta.yaml:43]:
```suggestion
    requirements:
      host:
        - cuda-version {{ cuda_version }}
      run:
        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
```

File: recipes/cuda-nvtx/meta.yaml
Comment (bdice) [inline recipes/cuda-nvtx/meta.yaml:58]:
```suggestion
    requirements:
      host:
        - cuda-version {{ cuda_version }}
      run:
        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
        - cuda-nvtx >={{ version }}
        - cuda-nvtx-dev_{{ target_platform }} =={{ version }}
```

File: recipes/cuda-nvtx/meta.yaml
Comment (bdice) [inline recipes/cuda-nvtx/meta.yaml:72]:
```suggestion
    requirements:
      host:
        - cuda-version {{ cuda_version }}
      run:
        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
```

File: recipes/cuda-nvtx/meta.yaml
Comment (bdice) [inline recipes/cuda-nvtx/meta.yaml:62]: Maybe change all `doc_url`s to https://docs.nvidia.com/nvtx/?
```suggestion
      doc_url: https://docs.nvidia.com/nvtx/
```

File: recipes/cuda-nvtx/meta.yaml
Comment (bdice) [inline recipes/cuda-nvtx/meta.yaml:105]:
```suggestion
      doc_url: https://docs.nvidia.com/nvtx/
```

File: recipes/cuda-nvtx/meta.yaml
Comment (bdice) [inline recipes/cuda-nvtx/meta.yaml:117]:
```suggestion
  doc_url: https://docs.nvidia.com/nvtx/
```

**Review by bdice — COMMENTED** [review]
Recipe looks fine aside from minor suggestions on `doc_url`. I haven't checked the package contents yet.

**Comment by jakirkham** [comment]
Currently we have [`nvtx-c`]( https://github.com/conda-forge/nvtx-c-feedstock ), which packages as similar library based on the [NVTX repo]( https://github.com/NVIDIA/NVTX/ )'s [C source code]( https://github.com/NVIDIA/NVTX/tree/release-v3/c ).

This in turn is used by [`nvtx`, Python bindings to the C library]( https://github.com/conda-forge/nvtx-feedstock ), which is based on [this Python code]( https://github.com/NVIDIA/NVTX/tree/release-v3/python ).

With this PR, we would now have `cuda-nvtx`, which packages NVTX based on the CTK copy.

This situation is somewhat reminiscent of the CCCL situation we just worked through

Wonder if we should do something to...

1. resolve conflicts
2. disambiguate packages
3. sunset unused components
4. ?

Would be interested in hearing thoughts from folks here 🙂

**Comment by jakirkham** [comment]
cc @shwina (who may also have thoughts on this)

**Comment by shwina** [comment]
We could change the `nvtx` Python package to depend on `cuda-nvtx`?

**Comment by bdice** [comment]
I support @shwina’s plan.

**Comment by jakirkham** [comment]
Generally that seems reasonable. One suggestion would be to introduce a conflict in `nvtx-c` analogous to what we did for `cub` ( https://github.com/conda-forge/cub-feedstock/pull/23 )

File: recipes/cuda-nvtx/bld.bat
Comment (vyasr) [inline recipes/cuda-nvtx/bld.bat:3]: This is missing libraries.

Not sure if we also want to include bin for completeness. IMO we've been doing that in all packages, even those without binaries, so we may as well stay consistent, but it's not a big deal.

**Review by vyasr — CHANGES_REQUESTED** [review]
One change request.

Also, do we need to make changes to nvtx-c before we merge this? CC @jakirkham

File: recipes/cuda-nvtx/bld.bat
Comment (adibbley) [reply recipes/cuda-nvtx/bld.bat:3]: There are no libraries in the Windows source. 

Which components have bin when it doesn't exist? This script would fail in that case.

File: recipes/cuda-nvtx/meta.yaml
Comment (vyasr) [inline recipes/cuda-nvtx/meta.yaml:96]: Should we link to https://docs.nvidia.com/nvtx/ instead? The docs may be the best source, not sure if there is an official nvtx page. Maybe the [Github source](https://github.com/NVIDIA/NVTX) is better.

Comment (adibbley) [reply recipes/cuda-nvtx/meta.yaml:96]: I think the cuda-toolkit url makes sense for `home` here. We already link to their docs and this will only be the `nvtx` bundled with the CTK.

File: recipes/cuda-nvtx/bld.bat
Comment (vyasr) [reply recipes/cuda-nvtx/bld.bat:3]: Hmm [libcurand](https://github.com/conda-forge/libcurand-feedstock/blob/main/recipe/bld.bat) and [libcublas](https://github.com/conda-forge/libcublas-feedstock/blob/main/recipe/bld.bat), for instance. Maybe this is a case where we haven't been careful enough in our Windows scripts since we're not building Windows packages atm?

File: recipes/cuda-nvtx/meta.yaml
Comment (vyasr) [reply recipes/cuda-nvtx/meta.yaml:96]: Fine with me.

**Review by vyasr — APPROVED** [review]
Approving assuming the nvtx-c conflict is addressed at some point.

File: recipes/cuda-nvtx/meta.yaml
Comment (shwina) [reply recipes/cuda-nvtx/meta.yaml:96]: Apologies: https://nvtx.readthedocs.io/en/latest/

Comment (bdice) [reply recipes/cuda-nvtx/meta.yaml:96]: @shwina That link is for the Python package -- we're not shipping that here. This package is C++ headers/libraries only, so I think the other proposals (https://developer.nvidia.com/cuda-toolkit or https://docs.nvidia.com/nvtx/ or https://github.com/NVIDIA/NVTX) would be preferred.

**Review by bdice — APPROVED** [review]

**Comment by kkraus14** [comment]
I think we need a repodata patch on nvtx-c to ensure these don't clobber each other and then we can maybe consider making it an alias?

Additionally, what is the relationship and versioning in relation to the GitHub repo https://github.com/NVIDIA/NVTX?

File: recipes/cuda-nvtx/bld.bat
Comment (jakirkham) [inline recipes/cuda-nvtx/bld.bat:3]:
```suggestion
move include\nvtx3 %LIBRARY_INC%\targets\x64
```

File: recipes/cuda-nvtx/meta.yaml
Comment (jakirkham) [inline recipes/cuda-nvtx/meta.yaml:76]:
```suggestion
      - Library\include\targets\{{ target_name }}  # [win]
```

File: recipes/cuda-nvtx/meta.yaml
Comment (jakirkham) [inline recipes/cuda-nvtx/meta.yaml:94]:
```suggestion
        - test -L $PREFIX/lib/libnvToolsExt.so                                              # [linux]
        - test -f $PREFIX/lib/pkgconfig/nvToolsExt*.pc                                      # [linux]
        - test -f $PREFIX/targets/{{ target_name }}/include/nvToolsExtCuda.h                # [linux]
        - test -f $PREFIX/targets/{{ target_name }}/include/nvtx3/nvToolsExtCuda.h          # [linux]
        - test -L $PREFIX/targets/{{ target_name }}/lib/libnvToolsExt.so                    # [linux]
        - if not exist Library\include\targets\{{ target_name }}\nvtx3\nvToolsExt.h exit 1  # [win]
```

**Review by jakirkham — COMMENTED** [review]
To avoid clobbering `nvtx-c`, suggesting that we place the contents here in `targets` on Windows as we did with `cuda-cccl` (for a similar reason)

File: recipes/cuda-nvtx/meta.yaml
Comment (jakirkham) [inline recipes/cuda-nvtx/meta.yaml:38]:
```suggestion
  - name: cuda-nvtx
    # `cuda-nvtx` does not ship `*.dll`s on Windows. So this is only relevant on Linux
    build:
       skip: True  # [not linux]
```

**Review by jakirkham — COMMENTED** [review]
Noticed a clobbering issue regarding headers on [Windows CI]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=691590&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=2fb5b0a7-737c-5f5c-8f3f-f6db6174bacf&l=1345 )

In looking into this, it appears `cuda-nvtx` may not actually be needed on Windows since there are no `.dll`s in the `redist` or the package here and the headers are already in `cuda-nvtx-dev`

Suggesting we disable `cuda-nvtx` builds on Windows

File: recipes/cuda-nvtx/meta.yaml
Comment (jakirkham) [inline recipes/cuda-nvtx/meta.yaml:70]: Also if we drop `cuda-nvtx` on Windows, this would need to change

```suggestion
    build:                                                 # [linux]
      run_exports:                                         # [linux]
        - {{ pin_subpackage("cuda-nvtx", max_pin='x') }}   # [linux]
```

File: recipes/cuda-nvtx/meta.yaml
Comment (jakirkham) [inline recipes/cuda-nvtx/meta.yaml:84]:
```suggestion
        - {{ pin_subpackage("cuda-nvtx", exact=True) }}  # [win]
```

**Review by jakirkham — COMMENTED** [review]
If we drop `cuda-nvtx` on Windows, there would be some follow up changes needed. Added those below

File: recipes/cuda-nvtx/meta.yaml
Comment (jakirkham) [reply recipes/cuda-nvtx/meta.yaml:96]: Let's stick with what we have for now. If we decide this needs to be changed, we can open an issue with the proposed change and more context and then follow up from there

File: recipes/cuda-nvtx/meta.yaml
Comment (jakirkham) [inline recipes/cuda-nvtx/meta.yaml:41]: Maybe this is a better approach for the clobbering issue on Windows?

```suggestion
    files: []                                 # [win]
    files:                                    # [linux]
      - lib/libnv*.so.*                       # [linux]
      - targets/{{ target_name }}/lib/*.so.*  # [linux]
```

File: recipes/cuda-nvtx/meta.yaml
Comment (jakirkham) [inline recipes/cuda-nvtx/meta.yaml:40]: To fix the linter error

```suggestion
        skip: true  # [not linux]
```

**Comment by jakirkham** [comment]
Resolved the Windows clobbering issues by dropping `cuda-nvtx`. That package was not needed on Windows anyways (no runtime libraries & headers in `cuda-nvtx-dev`).

**Comment by jakirkham** [comment]
Thanks all! 🙏

Let's follow up on anything else in the feedstock 🙂

**Pull request merged**
